File: src/main/java/com/adven/concordion/extensions/exam/files/commands/FilesSetCommand.java
package com.adven.concordion.extensions.exam.files.commands;

import com.adven.concordion.extensions.exam.core.html.Html;
import com.adven.concordion.extensions.exam.files.FilesLoader;
import org.concordion.api.CommandCall;
import org.concordion.api.Evaluator;
import org.concordion.api.ResultRecorder;

import static com.adven.concordion.extensions.exam.core.html.HtmlBuilder.*;
import static java.io.File.separator;
import static kotlin.TuplesKt.to;

public class FilesSetCommand extends BaseCommand {
    private FilesLoader filesLoader;

    public FilesSetCommand(String name, String tag, FilesLoader filesLoader) {
        super(name, tag);
        this.filesLoader = filesLoader;
    }

    @Override
    public void setUp(CommandCall commandCall, Evaluator evaluator, ResultRecorder resultRecorder) {
        Html root = tableSlim(commandCall.getElement());

        final String path = root.takeAwayAttr("dir");
        if (path != null) {

            String evalPath = evaluator.evaluate(path).toString();

            filesLoader.clearFolder(evalPath);

            root.childs(flCaption(evalPath));
            addHeader(root, HEADER, FILE_CONTENT);
            boolean empty = true;
            for (Html f : root.childs()) {
                if ("file".equals(f.localName())) {
                    final FilesLoader.FileTag fileTag = filesLoader.readFileTag(f, evaluator);
                    filesLoader.createFileWith(evalPath + separator + fileTag.getName(), fileTag.getContent());
                    root.childs(
                        trWithTDs(
                            span(fileTag.getName()),
                            codeXml(fileTag.getContent()).attrs(
                                to("autoFormat", String.valueOf(fileTag.getAutoFormat())),
                                to("lineNumbers", String.valueOf(fileTag.getLineNumbers()))
                            ))).remove(f);

                    empty = false;
                }
            }
            if (empty) {
                addRow(root, EMPTY, "");
            }
        }
    }
}

File: exam-files/src/main/java/io/github/adven27/concordion/extensions/exam/files/commands/FilesSetCommand.kt
package io.github.adven27.concordion.extensions.exam.files.commands

import io.github.adven27.concordion.extensions.exam.core.content
import io.github.adven27.concordion.extensions.exam.core.html.codeXml
import io.github.adven27.concordion.extensions.exam.core.html.span
import io.github.adven27.concordion.extensions.exam.core.html.table
import io.github.adven27.concordion.extensions.exam.core.html.trWithTDs
import io.github.adven27.concordion.extensions.exam.files.FileTester
import java.io.File
import org.concordion.api.CommandCall
import org.concordion.api.Evaluator
import org.concordion.api.Fixture
import org.concordion.api.ResultRecorder

class FilesSetCommand(private val fileTester: FileTester) : BaseCommand<Any?, Any?>(setOf()) {
    override fun model(context: Context): Any? {
        TODO("Not yet implemented")
    }

    override fun render(commandCall: CommandCall, result: Any?) {
        TODO("Not yet implemented")
    }

    override fun process(model: Any?, eval: Evaluator, recorder: ResultRecorder): Any? {
        TODO("Not yet implemented")
    }

    override fun setUp(
        commandCall: CommandCall,
        evaluator: Evaluator,
        resultRecorder: ResultRecorder,
        fixture: Fixture
    ) {
        val root = table(commandCall.element)
        val path = root.getAttr("dir")
        if (path != null) {
            val evalPath = evaluator.evaluate(path).toString()
            fileTester.clearFolder(evalPath)
            root.invoke(flCaption(evalPath))
            addHeader(root, HEADER, FILE_CONTENT)
            var empty = true
            for (f in root.childs()) {
                if ("file" == f.localName()) {
                    f.content()
                    val file = f.content(evaluator)
                    fileTester.createFile(evalPath + File.separator + f.attr("name"), file)
                    root.invoke(
                        trWithTDs(
                            span(f.attr("name")),
                            codeXml(file)
                        )
                    ).remove(f)
                    empty = false
                }
            }
            if (empty) {
                addRow(root, EMPTY, "")
            }
        }
    }
}
